controllers/storagecluster/consumer.go

```diff
@@ -0,0 +1,146 @@
+		return nil
+	}
+
+	if instance.Status.ExternalStorage.ConsumerID == "" {
+		return onBoardConsumer(r, instance)
+	}
+
+	if !instance.Status.ExternalStorage.GrantedCapacity.Equal(*instance.Spec.ExternalStorage.RequestedCapacity) {
+		return upscaleOrDownscaleConsumer(r, instance)
+	}
+
+	return nil
+}
+
+func (o *ocsConsumer) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+	if !(instance.Spec.ExternalStorage.Enable && instance.Spec.ExternalStorage.StorageProviderKind == ocsv1.KindOCS) {
+		return nil
+	}
+
+	return offBoardConsumer(r, instance)
+}
+
+func onBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	consumerClient, err := getConsumerClient(instance)
+	if err != nil {
+		return err
+	}
+
+	capacity := func(instance *ocsv1.StorageCluster) string {
+		if instance.Spec.ExternalStorage.RequestedCapacity == nil {
+			q := resource.MustParse("1Ti")
+			instance.Spec.ExternalStorage.RequestedCapacity = &q
+		}
+		return instance.Spec.ExternalStorage.RequestedCapacity.String()
+	}(instance)
+
+	// TODO: think about naming convention
+	response, err := consumerClient.OnboardConsumer("my-ocs", capacity)
+	if err != nil {
+		requeue := r.handleOnBoardConsumerErrors(instance, err)
+		if !requeue {
+			return nil
+		}
+		return err
+	}
+
+	if response.StorageConsumerUUID == "" || response.GrantedCapacity == "" {
+		return fmt.Errorf("response is empty")
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return nil
+}
+
+func offBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	return nil
+}
+
+func upscaleOrDownscaleConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	return nil
+}
+
+func getExternalConfigFromProvider(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) ([]ExternalResource, error) {
+
+	if externalOCSResources[instance.UID] != nil {
+		return externalOCSResources[instance.UID], nil
+	}
+
+	consumerClient, err := getConsumerClient(instance)
+	if err != nil {
+		return nil, err
+	}
+
+	consumerID := instance.Status.ExternalStorage.ConsumerID
+
+	config, err := consumerClient.GetStorageConfig(consumerID)
+	if err != nil {
+		return nil, err
+	}
+
+	externalResources, err := parseExternalOcsJSONData(config.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	externalOCSResources[instance.UID] = externalResources
+
+	return externalOCSResources[instance.UID], nil
+}
+
+func getExternalConfig(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) ([]ExternalResource, error) {
+	if instance.Spec.ExternalStorage.StorageProviderKind == ocsv1.KindOCS {
+		return getExternalConfigFromProvider(r, instance)
+	}
+	return r.retrieveExternalSecretData(instance)
+}
+
+func parseExternalOcsJSONData(jsonBlob []byte) ([]ExternalResource, error) {
```

Comment:
As it is a short fun and only used once, lets move the code there

Reply:
done